File: Formularios/Agregar/add_cobroRetencion.cs
using System;
using System.Data;
using System.Linq;
using System.Windows.Forms;
using System.Xml.Linq;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using Centrex.Models;

namespace Centrex
{

    public partial class add_cobroRetencion
    {
   private Form formViejo;
        private CentrexDbContext _context;

        // Parámetros de operación
        public bool Edicion { get; set; } = false;
        public bool Borrado { get; set; } = false;
        public TmpCobroRetencionEntity? EditaCobroRetencion { get; set; }

      public add_cobroRetencion()
        {
    InitializeComponent();
        }

      private void add_cobroRetencion_Load(object sender, EventArgs e)
     {
      _context = new CentrexDbContext();

            // Cargar combo de retenciones (Impuestos marcados como retención)
            cargar_comboRetenciones();

            if (!Edicion && !Borrado)
     {
    cmb_retencion.SelectedItem = null;
   cmb_retencion.Text = "Seleccione una retención...";
       }
    else if (EditaCobroRetencion is not null)
    {
        cmb_retencion.SelectedValue = EditaCobroRetencion.IdImpuesto;
  txt_importe.Text = EditaCobroRetencion.Total.ToString("0.00");
      txt_notas.Text = EditaCobroRetencion.Notas;
  }

  if (Borrado)
     {
 cmb_retencion.Enabled = false;
     txt_importe.Enabled = false;
        txt_notas.Enabled = false;
     cmd_ok.Enabled = false;
    cmd_exit.Enabled = false;
       Show();
              if (Interaction.MsgBox("¿Está seguro que desea borrar esta retención?", (MsgBoxStyle)((int)Constants.vbYesNo + (int)Constants.vbQuestion), "Centrex") == MsgBoxResult.Yes)
       {
  if (!cobros_retenciones.borrarTmpCobroRetencion(EditaCobroRetencion.IdTmpRetencion))
       {
   Interaction.MsgBox("No se ha podido borrar la retención.");
}
             }
        closeandupdate(this);
      }

  // formViejo = Form ' Comentado para evitar error de compilación
       // Form = Me ' Comentado para evitar error de compilación
      }

        private void cargar_comboRetenciones()
        {
     try
   {
                var lista = _context.ImpuestoEntity
       .Where(i => i.EsRetencion == true)
        .OrderBy(i => i.Nombre)
         .Select(i => new { Id = i.IdImpuesto, Nombre = i.Nombre })
        .ToList();

      cmb_retencion.DataSource = lista;
     cmb_retencion.DisplayMember = "Nombre";
    cmb_retencion.ValueMember = "Id";
          }
         catch (Exception ex)
      {
        Interaction.MsgBox("Error al cargar las retenciones: " + ex.Message, Constants.vbExclamation);
    }
        }

        private void txt_importe_KeyPress(object sender, KeyPressEventArgs e)
     {
  // Solo permitir números y punto decimal
            if (!(char.IsDigit(e.KeyChar) || e.KeyChar == '\b' || e.KeyChar == '.'))
            {
     e.Handled = true;
            }
        }

        private void cmd_exit_Click(object sender, EventArgs e)
        {
            closeandupdate(this);
        }

        private void cmd_ok_Click(object sender, EventArgs e)
        {
   if (cmb_retencion.SelectedItem is null)
            {
            Interaction.MsgBox("Debe seleccionar una retención", (MsgBoxStyle)((int)Constants.vbExclamation + (int)Constants.vbOKOnly), "Centrex");
         return;
   }
            else if (string.IsNullOrWhiteSpace(txt_importe.Text))
         {
    Interaction.MsgBox("Debe ingresar un importe", (MsgBoxStyle)((int)Constants.vbExclamation + (int)Constants.vbOKOnly), "Centrex");
   return;
            }

     var cb = new TmpCobroRetencionEntity()
 {
       IdImpuesto = Conversions.ToInteger(cmb_retencion.SelectedValue),
           Total = decimal.Parse(txt_importe.Text),
      Notas = txt_notas.Text
    };

            try
            {
       if (Edicion)
      {
        cb.IdTmpRetencion = EditaCobroRetencion.IdTmpRetencion;
      cobros_retenciones.updateTmpCobroRetencion(cb);
              }
        else
        {
         cobros_retenciones.addTmpCobroRetencion(cb);
                }
   }
        catch (Exception ex)
  {
         Interaction.MsgBox("Ocurrió un error al guardar la retención: " + ex.Message, (MsgBoxStyle)((int)Constants.vbExclamation + (int)Constants.vbOKOnly), "Centrex");
       }

    closeandupdate(this);
    }

     private void psearch_retencion_Click(object sender, EventArgs e)
        {
   // Reutiliza tu formulario de búsqueda general
     string tmpTabla = VariablesGlobales.tabla;
   VariablesGlobales.tabla = "retenciones";
  var frmSearch = new search("retenciones");
      frmSearch.ShowDialog();
    VariablesGlobales.tabla = tmpTabla;

       if (frmSearch.SelectedIndex > 0)
            {
   cmb_retencion.SelectedValue = frmSearch.SelectedIndex;
   }
      }

        private void add_cobroRetencion_FormClosed(object sender, FormClosedEventArgs e)
  {
          // Form = formViejo
 closeandupdate(this);
     }

        private void cmd_agregarImpuesto_Click(object sender, EventArgs e)
        {
            var frmImpuesto = new add_impuesto(true, false);
            frmImpuesto.ShowDialog();

     if (frmImpuesto.id_impuesto > 0)
   {
  cargar_comboRetenciones();
           cmb_retencion.SelectedValue = frmImpuesto.id_impuesto;
            }
   }
    }
}

File: __Viejo/Formularios/Agregar/add_cobroRetencion.vb
Public Class add_cobroRetencion
    Private formViejo As Form

    Private Sub add_cobroRetencion_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        cargar_combo(cmb_retencion, "SELECT id_impuesto, nombre FROM impuestos WHERE esRetencion = '1' ORDER BY nombre ASC", basedb, "nombre", "id_impuesto")

        If Not edicion And Not borrado Then
            cmb_retencion.SelectedItem = Nothing
            cmb_retencion.Text = "Seleccione una retención..."
        Else
            cmb_retencion.SelectedValue = edita_cobroRetencion.id_retencion

            txt_importe.Text = edita_cobroRetencion.total
            txt_notas.Text = edita_cobroRetencion.notas
        End If

        If borrado Then
            cmb_retencion.Enabled = False
            txt_importe.Enabled = False
            txt_notas.Enabled = False
            cmd_ok.Enabled = False
            cmd_exit.Enabled = False
            Me.Show()
            If MsgBox("¿Está seguro que desea borrar esta retención?", vbYesNo + vbQuestion, "Centrex") = MsgBoxResult.Yes Then
                If borrarTmptransferencia(edita_cobroRetencion.id_retencion) = False Then
                    MsgBox("No se ha podido borrar la retención.")
                End If
            End If
            closeandupdate(Me)
        End If

        formViejo = form
        form = Me
    End Sub

    Private Sub txt_importe_KeyPress(sender As Object, e As KeyPressEventArgs) Handles txt_importe.KeyPress
        e.Handled = valida(e.KeyChar, 5)
    End Sub

    Private Sub cmd_exit_Click(sender As Object, e As EventArgs) Handles cmd_exit.Click
        closeandupdate(Me)
    End Sub

    Private Sub cmd_ok_Click(sender As Object, e As EventArgs) Handles cmd_ok.Click
        Dim cb As New cobroRetencion

        If cmb_retencion.Text = "Seleccione una retencion..." Then
            MsgBox("Debe seleccionar una retención", vbExclamation + vbOKOnly, "Centrex")
            Exit Sub
        ElseIf txt_importe.Text = "" Then
            MsgBox("Debe ingresar un importe", vbExclamation + vbOKOnly, "Centrex")
            Exit Sub
        End If

        With cb
            .id_impuesto = cmb_retencion.SelectedValue
            .total = txt_importe.Text
            .notas = txt_notas.Text
        End With

        If edicion Then
            cb.id_retencion = edita_cobroRetencion.id_retencion
            If Not updateTmpCobroRetencion(cb) Then
                MsgBox("Ocurrió un error al actualizar la retención.", vbExclamation + vbOKOnly, "Centrex")
            End If
        Else
            cb.id_retencion = addTmpCobroRetencion(cb)
            If cb.id_retencion = False Then
                MsgBox("Ocurrió un error al agregar la retencion", vbExclamation + vbOKOnly, "Centrex")
            End If
        End If

        closeandupdate(Me)
    End Sub

    Private Sub psearch_retencion_Click(sender As Object, e As EventArgs) Handles psearch_retencion.Click
        Dim tmp As String
        tmp = tabla
        tabla = "retenciones"
        Me.Enabled = False
        search.ShowDialog()
        tabla = tmp

        'Establezco la opción del combo
        cmb_retencion.SelectedValue = id
    End Sub

    Private Sub add_cobroRetencion_FormClosed(sender As Object, e As FormClosedEventArgs) Handles Me.FormClosed
        form = formViejo
        closeandupdate(Me)
    End Sub

    Private Sub cmb_retencion_KeyPress(sender As Object, e As KeyPressEventArgs) Handles cmb_retencion.KeyPress
        'e.KeyChar = ""
    End Sub

    Private Sub cmd_agregarImpuesto_Click(sender As Object, e As EventArgs) Handles cmd_agregarImpuesto.Click
        Dim frmImpuesto As New add_impuesto(True, False)
        Dim i As Integer
        frmImpuesto.ShowDialog()
        i = frmImpuesto.id_impuesto

        cargar_combo(cmb_retencion, "SELECT id_impuesto, nombre FROM impuestos WHERE esRetencion = '1' ORDER BY nombre ASC", basedb, "nombre", "id_impuesto")
        cmb_retencion.SelectedValue = i
    End Sub
End Class